pkg/service/service.go

```diff
@@ -33,12 +33,6 @@ import (
 
 const anyPort = "*"
 
-var (
-	updateMetric = metrics.ServicesCount.WithLabelValues("update")
```

Comment:
This code is called before `NewLegacyMetrics` initializes ServicesCount, and calls `NoOpCounterVec.WithLabelValues` (returns `NoOpGauge`). That's the reason `services_events_total` is missing. Is my understanding correct?

Reply:
Yes, I think you're right. I opened a small thread about this in #development and that's what Andre suspected as well. :+1:

Reply:
cc @chancez

Reply:
Gotcha. How can we prevent this from reoccurring?

Reply:
I'm not too sure to be honest. We have github.com/cilium/customvet where we can write our own linting checkers. However, this is the only case that I've found in the code base. Moving forward there is an effort to move metrics into their modules (see https://github.com/cilium/cilium/pull/25651) which should help, but techincally not a guarantee.

Reply:
For example, we can configure a workflow for the metrics check in which we set up Prometheus and scrape the agent's metrics, fetch the metrics key list using promql-cli or Prometheus API, and then check if all metrics are available. But I'm not sure if such an expensive test suite is worth the cost.

Reply:
In the past, I've seen metrics not published unless Cilium modified its value. At least, this is what I observed with `cilium metrics list` and not sure if the same would apply to simply hitting the Prometheus endpoint. If they are not published, then this will make this workflow suggested a bit more complicated.

Reply:
The option is to get rid of these NoOp metrics potentially. The approach we're using is a bit odd from my perspective but I haven't really dug into why we do it this way. There's a lot of history probably in how we decided that approach.

We could choose to always the 'real' metrics, and then change the logic to not register them, but it's probably a bigger/harder change to make (and shouldn't block this PR). Then the metrics would still exist (they're just atomic counters) but they won't be exported (because they aren't registered).

Reply:
That seams quite reasonable to me. I guess I didn't question the structure, but now that I think about it a bit, they're likely defined this way (NoOp) because of the problem of having global metrics defined in one place rather than per module. Due to them being exposed as global vars, they can be accessed at any time without any guarantee that they've been initialized, so NoOp wrapper was created to prevent panics / crashes in case that happened. That's my guess.

cc @aanm (for adding NoOp) and @dylandreimerink (for whether we still need the idea of "NoOp" metrics going forward)

Reply:
> The option is to get rid of these NoOp metrics potentially. The approach we're using is a bit odd from my perspective but I haven't really dug into why we do it this way

The NoOp metrics exist as a stopgap, currently a lot of the code base still uses global variables to get metrics which is subject to ordering. 

> they're likely defined this way (NoOp) because of the problem of having global metrics defined in one place rather than per module. Due to them being exposed as global vars, they can be accessed at any time without any guarantee that they've been initialized, so NoOp wrapper was created to prevent panics / crashes in case that happened.

Yes, that is indeed correct.

Ideally, all metrics are provided to components via parameters during construction, but this conversion takes some time. We can remove NoOps once all global metric variables are gone.

Reply:
There's an intermediate step I mentioned that wasn't addressed: You can have globals, without NoOp metrics, but not call `registry.Register(yourMetric)`. This avoids ordering issues and nil globals.

Reply:
Btw, I've updated the second commit's description to describe the global variable vs initialization ordering as the root cause of the bug.

Reply:
> There's an intermediate step I mentioned that wasn't addressed: You can have globals, without NoOp metrics, but not call registry.Register(yourMetric). This avoids ordering issues and nil globals.

Yes, you could init the global var and then upon creating the registry determine if you want to add that metric to the registry or not. 

I am not sure why NoOps were chosen over that approach, it is not something invented for the modular metrics, it was already the way things worked. The main thing that changed was init ordering.

I would argue that removing all global variables in favor of dependency injection ASAP is a better use of cycles then refactoring the NoOps away first.